contracts/account/paymaster/PaymasterERC20.sol

```diff
@@ -77,27 +63,19 @@ abstract contract PaymasterERC20 is PaymasterCore {
         uint256 actualGasCost,
         uint256 actualUserOpFeePerGas
     ) internal virtual override {
-        bytes32 userOpHash = bytes32(context[0x00:0x20]);
-        IERC20 token = IERC20(address(bytes20(context[0x20:0x34])));
-        uint256 prefundAmount = uint256(bytes32(context[0x34:0x54]));
-        uint256 tokenPrice = uint256(bytes32(context[0x54:0x74]));
-        address user = address(bytes20(context[0x74:0x88]));
-        address guarantor = address(bytes20(context[0x88:0x9C]));
-
-        uint256 actualAmount = (actualGasCost +
-            (guarantor == address(0)).ternary(POST_OP_COST, POST_OP_COST_WITH_GUARANTOR) *
-            actualUserOpFeePerGas).mulDiv(tokenPrice, _tokenPriceDenominator());
-
-        if (guarantor == address(0)) {
-            token.safeTransfer(user, prefundAmount - actualAmount);
-            emit UserOperationSponsored(userOpHash, user, address(0), actualAmount, tokenPrice, false);
-        } else if (token.trySafeTransferFrom(user, address(this), actualAmount)) {
-            token.safeTransfer(guarantor, prefundAmount);
-            emit UserOperationSponsored(userOpHash, user, guarantor, actualAmount, tokenPrice, false);
-        } else {
-            token.safeTransfer(guarantor, prefundAmount - actualAmount);
-            emit UserOperationSponsored(userOpHash, user, guarantor, actualAmount, tokenPrice, true);
-        }
+        (
+            bytes32 userOpHash,
+            IERC20 token,
+            uint256 prefundAmount,
+            uint256 tokenPrice,
+            address userOpSender /* address prefundPayer */,
+
+        ) = _decodeContext(context);
+
+        uint256 actualAmount = _erc20Cost(actualGasCost, actualUserOpFeePerGas, tokenPrice);
+        token.safeTransfer(userOpSender, prefundAmount - actualAmount);
```

Comment:
if `_prefundPayer` is overriden, and the returned value is NOT `userOpSender`, then this contract will take the prefund from that account and give the refund to the `userOpSender`

This incorrect.